### PR TITLE
Frontend with vitejs

### PR DESCRIPTION
**Pull request by thisconnect**
For more information see https://vitejs.dev/

This PR also updates React to version 18

### CODE REVIEW

**Comment by thisconnect** [comment]
`make buildweb` with this PR (11e2191769f7b1124bddb6d1f2dc95f85b38e4ad)

    build/assets/qrreader-9228c72e.js                               291.60 kB │ gzip:  88.50 kB
    build/assets/index-d1fa6013.js                                1,882.36 kB │ gzip: 633.40 kB

`make buildweb` with this master (7afc6f8570c263606ccd3173c4cd97e870575ba4)

    644.14 kB  build/static/js/main.06fd36e9.js
    87.22 kB   build/static/js/459.4bf01830.chunk.js

File: Makefile [hunk 1]
```diff
@@ -53,7 +53,7 @@ buildweb:
 	node --version
 	npm --version
 	rm -rf ${WEBROOT}/build
-	cd ${WEBROOT} && npm ci --ignore-scripts
+	cd ${WEBROOT} && npm ci
```
Comment (thisconnect) [inline Makefile:56]: with `--ignore-scripts` esbuild can't get the correct binary for linux and the build failed

File: frontends/web/src/utils/env.ts [hunk 2]
```diff
@@ -15,7 +15,7 @@
  * limitations under the License.
  */
 
-export const debug = process.env.NODE_ENV === 'development';
+export const debug = import.meta.env.DEV;
```
Comment (thisconnect) [inline frontends/web/src/utils/env.ts:18]: is `true` in `make webdev` and `false` in the Qt app.

File: frontends/web/src/utils/qwebchannel.js [hunk 3]
```diff
@@ -418,10 +418,3 @@ function QObject(name, data, webChannel)
         object[name] = data.enums[name];
     }
 }
-
-//required for use with nodejs
-if (typeof module === 'object') {
-    module.exports = {
-        QWebChannel: QWebChannel
-    };
-}
```
Comment (thisconnect) [inline frontends/web/src/utils/qwebchannel.js:427]: byebye (last?) commonjs module, welcome es6 module 🎉

File: frontends/web/vite.config.mjs [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    },
+    plugins: [
+      react(),
+      eslint(),
+      checker({
+        typescript: true,
+      }),
+    ],
+    test: {
+      css: false,
```
Comment (thisconnect) [inline frontends/web/vite.config.mjs:20]: tests happily passed with css set to `false` here.

**Comment by thisconnect** [comment]
did a quick test and merged WC and qr-scanner 
- https://github.com/thisconnect/bitbox-wallet-app/tree/frontend-vitejs-with-wc
- https://github.com/thisconnect/bitbox-wallet-app/tree/frontend-vitejs-with-qrscanner

qrsscanner seems ok (loads video element with scanner) and WC builds + webdev runs now without weird warnings.

**Review by thisconnect — COMMENTED** [review]
thanks @shonsirsha for your help for migrating the tests.

**Comment by thisconnect** [comment]
We should test this on all platforms, to get the new dependencies installed please run `make buildweb`.

**Comment by thisconnect** [comment]
@Beerosagos could you check this PR if it works on your machine?

**Review by Beerosagos — APPROVED** [review]
Didn't look much into the code, but tACK on linux and android :+1: amazing job :)

File: frontends/web/index.html [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+    <meta charset="utf-8" />
+    <meta name="viewport" content="width=device-width, initial-scale=1" />
+    <meta name="theme-color" content="#000000" />
+    <title>BitBoxApp</title>
+  </head>
+  <body>
+    <div id="root"></div>
+    <script type="module" src="/src/index.tsx"></script>
+  </body>
+</html>
```
[the remark this answers is not in the document]
Comment (thisconnect) [reply frontends/web/index.html:13]: nope, I moved it, vitejs expects it in the root and not in src nor public.

https://vitejs.dev/guide/#index-html-and-project-root

**Comment by thisconnect** [comment]
also @benma @jadzeidan PTAL and test `make buildweb` and `make webdev` to confirm that this works on your OS.

**Comment by thisconnect** [comment]
rebased, but butchered `frontends/web/src/hooks/qrcodescanner.test.ts` 😞

File: frontends/web/src/components/forms/input.test.tsx [hunk 6]
```diff
@@ -1,6 +1,6 @@
 /**
  * Copyright 2018 Shift Devices AG
- * Copyright 2021 Shift Crypto AG
+ * Copyright 2023 Shift Crypto AG
```
Comment (benma) [inline frontends/web/src/components/forms/input.test.tsx:3]: generally one does not delete old copyrights. just extend using "2021, 2023" (or 2021-2023 if 2022 also had changes)

File: frontends/web/src/hooks/qrcodescanner.test.ts [hunk 7]
```diff
@@ -13,37 +13,40 @@
...
-jest.mock('qr-scanner', () => {
+vi.mock('qr-scanner', () => {
   class MockedQrScanner {
     start = mockScanner.start();
     stop = mockScanner.stop();
     destroy = mockScanner.destroy();
-    static hasCamera: jest.Mock<Promise<void>, []>;
+    // TODO: I don't know how to migrate this
+    // @ts-ignore
+    static hasCamera: vi.Mock<Promise<void>, []>;
```
Comment (thisconnect) [inline frontends/web/src/hooks/qrcodescanner.test.ts:37]: this is blocking this PR  and the only issue that I am aware of. I am thinking about just nuking the whole test.

File: frontends/web/src/components/forms/input.test.tsx (hunk 6 above)
Comment (thisconnect) [reply frontends/web/src/components/forms/input.test.tsx:3]: Oups really? I don't recall we ever do that in bitbox-wallet-app

I'd prefer 2021-2023 as it makes more sense to me than `"2021, 2022, 2023, 2024` (I've never seen that anywhere)

Comment (thisconnect) [reply frontends/web/src/components/forms/input.test.tsx:3]: I checked all Copyright headers, nowhere we use multiple years.. I'll just undo the copyright changes in this PR.

File: frontends/web/src/hooks/qrcodescanner.test.ts (hunk 7 above)
Comment (thisconnect) [reply frontends/web/src/hooks/qrcodescanner.test.ts:37]: ack. I'll remove the test. I feel a bit sorry to kill it 🙈

**Comment by thisconnect** [comment]
rebased

**Comment by benma** [comment]
@thisconnect I am seeing this msg when building

```
One of your dependencies, babel-preset-react-app, is importing the                                                                                                
"@babel/plugin-proposal-private-property-in-object" package without                                                                                                                                                declaring it in its dependencies. This is currently working because                                                                                                                                                "@babel/plugin-proposal-private-property-in-object" is already in your                                                                                                                                             node_modules folder for unrelated reasons, but it may break at any time.             
                                                                                                                                                                                                                                                                                                                                                 babel-preset-react-app is part of the create-react-app project, which                                                                                                                                              is not maintianed anymore. It is thus unlikely that this bug will                                                                                                                                                  ever be fixed. Add "@babel/plugin-proposal-private-property-in-object" to                                                                                                                                          your devDependencies to work around this error. This will make this message                                                                                                                                        go away.                                         
```

Maybe you can get rid of babel-preset-react-app or fix the warning?

**Review by benma — APPROVED** [review]
On Windows I get

```
$ make buildweb
node --version
v18.16.0
npm --version
9.6.5
rm -rf frontends/web/build
cd frontends/web && npm ci
npm WARN deprecated @babel/plugin-proposal-private-methods@7.18.6: This proposal has been merged to the ECMAScript standard and thus this plugin is no longer maintained. Please use @babel/plugin-transform-private-methods instead.
npm WARN deprecated @babel/plugin-proposal-numeric-separator@7.18.6: This proposal has been merged to the ECMAScript standard and thus this plugin is no longer maintained. Please use @babel/plugin-transform-numeric-separator instead.
npm WARN deprecated @babel/plugin-proposal-class-properties@7.18.6: This proposal has been merged to the ECMAScript standard and thus this plugin is no longer maintained. Please use @babel/plugin-transform-class-properties instead.
npm WARN deprecated @babel/plugin-proposal-nullish-coalescing-operator@7.18.6: This proposal has been merged to the ECMAScript standard and thus this plugin is no longer maintained. Please use @babel/plugin-transform-nullish-coalescing-operator instead.
npm WARN deprecated @babel/plugin-proposal-optional-chaining@7.21.0: This proposal has been merged to the ECMAScript standard and thus this plugin is no longer maintained. Please use @babel/plugin-transform-optional-chaining instead.

```

Apart fom this it worked as expected. tACK on linux and windows.

Very nice!

**Pull request merged**